src/libAtomVM/nifs.c

```diff
@@ -2069,18 +2142,50 @@ static term nif_erlang_atom_to_list_1(Context *ctx, int argc, term argv[])
         RAISE_ERROR(OUT_OF_MEMORY_ATOM);
     }
 
-    if (UNLIKELY(memory_ensure_free_opt(ctx, atom_len * 2, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+    atom_table_write_bytes(ctx->global->atom_table, atom_ref, atom_len, atom_buf);
+
+    size_t u8len = unicode_buf_utf8_len((uint8_t *) atom_buf, atom_len);
+    bool latin1 = atom_len == u8len;
+
+    size_t list_len = latin1 ? atom_len : u8len;
+
+    if (UNLIKELY(memory_ensure_free_opt(ctx, list_len * 2, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+        free(atom_buf);
         RAISE_ERROR(OUT_OF_MEMORY_ATOM);
     }
 
-    atom_table_write_bytes(ctx->global->atom_table, atom_ref, atom_len, atom_buf);
-
     term prev = term_nil();
-    for (int i = atom_len - 1; i >= 0; i--) {
-        char c = atom_buf[i];
-        prev = term_list_prepend(term_from_int11(c), prev, &ctx->heap);
-    }
 
+    if (latin1) {
+        for (int i = atom_len - 1; i >= 0; i--) {
+            char c = atom_buf[i];
+            prev = term_list_prepend(term_from_int11(c), prev, &ctx->heap);
+        }
+    } else {
+        uint32_t *codepoints = malloc(u8len * sizeof(uint32_t));
+        if (IS_NULL_PTR(codepoints)) {
+            free(atom_buf);
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        uint8_t *u_in = (uint8_t *) atom_buf;
+        for (size_t i = 0; i < u8len; i++) {
+            size_t codepoint_size;
+            enum UnicodeTransformDecodeResult result
+                = bitstring_utf8_decode(u_in, atom_len, &codepoints[i], &codepoint_size);
+            if (UNLIKELY((result != UnicodeTransformDecodeSuccess)
+                    || !unicode_is_valid_codepoint(codepoints[i]))) {
+                free(codepoints);
+                free(atom_buf);
+                RAISE_ERROR(BADARG_ATOM);
```

Comment:
Is this a `badarg` or more of an internal logic error?  For example, if the atom is in fact a UTF8 encoded atom, is it even possible for the user to supply an atom that would fail?

We don’t have the C++ equivalent of `logic_error`, but if this is, then maybe we should, or maybe we should abort here.

Can you think of a case in which we could ever hit this condition?